xla/service/float8_fnuz_ir_emitter.cc

```diff
@@ -0,0 +1,646 @@
+llvm::Value* LastMantissaBit(PrimitiveType input_type, llvm::Value* value,
+                             PrimitiveType output_type, llvm::IRBuilder<>* b) {
+  const int src_mantissa_bits = SignificandWidth(input_type) - 1;
+  const int dest_mantissa_bits = SignificandWidth(output_type) - 1;
+  llvm::Type* int_type = b->getIntNTy(BitWidth(input_type));
+
+  llvm::Value* exponent = ExtractExponent(input_type, value, b);
+
+  // The index when the input/output is normal.
+  llvm::Value* last_bit_index =
+      llvm::ConstantInt::get(int_type, src_mantissa_bits - dest_mantissa_bits);
+
+  // Increase the index if the output will be denormal given the exponent.
+  llvm::Value* denormal_shift = b->CreateSub(
+      llvm::ConstantInt::get(int_type, UnderflowExponent(output_type) - 1),
+      exponent);
+  denormal_shift = Max(int_type, denormal_shift, 0, b);
+  last_bit_index = b->CreateAdd(last_bit_index, denormal_shift);
+
+  // The log_2(x) of the smallest denorm value. This gives us the exponent n
+  // that produces that number. This corresponds to the encoding with a single
+  // bit set in the last significant bit (0b0000...01).
+  const int input_log_minimum =
+      UnderflowExponent(input_type) - SignificandWidth(input_type);
+  const int output_log_minimum =
+      UnderflowExponent(output_type) - SignificandWidth(output_type);
+
+  // Alternatively, the input might be a denorm. This directly computes the
+  // last mantissa bit when the input is a denorm.  Suppose we have 2.664E-5
+  // (0x1.BFp-16) as the input number. The bit encoding for this in F16 is:
+  //   S|EEEEE|MMMMMMMMMM
+  // 0b0|00000|0110111111 = 2^(-14) * 0b0.0110111111
+  //
+  // To cast this to F8E5M2FNUZ, we would find the smallest denorm encoding and
+  // find the corresponding bit in the input.
+  //   S|MMMMM|MM
+  // 0b0|00000|01 = 2^(-15) * 0b0.01
+  // We can see the "last" bit of an output denorm represents 2^(-17), so we
+  // find that corresponding bit in the input. In this F16 example it is
+  // highlighted here:
+  //   S|EEEEE|MMMMMMMMMM
+  // 0b0|00000|0110111111
+  //             ^-- last mantissa bit
+  llvm::Value* denorm_last_mantissa_bit =
+      llvm::ConstantInt::get(int_type, output_log_minimum - input_log_minimum);
+
+  // Select the last mantissa bit based on whether the input is a normal number.
+  llvm::Value* normal_pred = IsNormalNumber(input_type, value, b);
+
+  // For the purposes of this function, consider zero a normal number.
+  normal_pred = b->CreateOr(normal_pred, IsZero(input_type, value, b));
+
+  // Select the normal or denorm case.
+  llvm::Value* last_mantissa_bit =
+      b->CreateSelect(normal_pred, last_bit_index, denorm_last_mantissa_bit);
+
+  // Ensure the last_mantissa_bit is a valid bit in the input mantissa.
+  // last_mantissa_bit is allowed to be the least significant bit of the
+  // exponent.
+  return Min(int_type, last_mantissa_bit, src_mantissa_bits + 1, b);
```

Comment:
The least significant bit of the exponent is `src_mantissa_bits`, not `src_mantissa_bits + 1`, since the mantissa field consists of indices `[0, src_mantissa_bits - 1]`.